src/lib.rs

```diff
@@ -1065,14 +1065,71 @@ pub unsafe trait KnownLayout {
     where
         Self: Sized;
 
+    /// `()` for sized types and `usize` for slice DSTs.
+    #[doc(hidden)]
+    type PointerMetadata: PointerMetadata;
+
     #[doc(hidden)]
     const LAYOUT: DstLayout;
 
     /// SAFETY: The returned pointer has the same address and provenance as
     /// `bytes`. If `Self` is a DST, the returned pointer's referent has `elems`
-    /// elements in its trailing slice. If `Self` is sized, `elems` is ignored.
+    /// elements in its trailing slice.
     #[doc(hidden)]
-    fn raw_from_ptr_len(bytes: NonNull<u8>, elems: usize) -> NonNull<Self>;
+    fn raw_from_ptr_len(bytes: NonNull<u8>, meta: Self::PointerMetadata) -> NonNull<Self>;
+}
+
+/// The metadata associated with a [`KnownLayout`] type.
+#[doc(hidden)]
+pub trait PointerMetadata {
+    /// Constructs a `Self` from an element count.
+    ///
+    /// If `Self = ()`, this returns `()`. If `Self = usize`, this returns
+    /// `elems`. No other types are currently supported.
+    fn from_elem_count(elems: usize) -> Self;
+
+    /// What is the size of the object with the given layout and pointer
+    /// metadata?
+    ///
+    /// # Panics
+    ///
+    /// If `Self = ()`, `layout` must describe a sized type. If `Self = usize`,
+    /// `layout` must describe a slice DST. Otherwise, `size_for_metadata` will
+    /// panic.
+    fn size_for_metadata(&self, layout: DstLayout) -> Option<usize>;
+}
+
+impl PointerMetadata for () {
+    #[inline]
+    #[allow(clippy::unused_unit)]
+    fn from_elem_count(_elems: usize) -> () {}
+
+    #[inline]
+    fn size_for_metadata(&self, layout: DstLayout) -> Option<usize> {
+        match layout.size_info {
+            SizeInfo::Sized { size } => Some(size),
+            SizeInfo::SliceDst(_) => unreachable!(),
```

Comment:
In the interest of minimizing our panic paths, perhaps we should do something besides `unreachable!()`, like return `None`.

(We can't use `unreachable_unchecked()`, because this method is safe and this line is obviously not actually unreachable when evaluated in isolation.)

Reply:
Done.